File: irrgarten/Game.java
package irrgarten;

import java.util.ArrayList;

public class Game {
    private static final int MAX_ROUNDS = 10;
    private int currentPlayerIndex;
    private String log;
    private ArrayList<Monster> monsters;
    private ArrayList<Player> players;
    private Player currentPlayer = new Player('0',Dice.randomIntelligence(), Dice.randomStrength());
    private Labyrinth labyrinth = new Labyrinth(10,20,9,17);

    public Game(int nplayers){
        // Inicializar la lista de jugadores y monstruos
        players = new ArrayList<>();
        monsters = new ArrayList<>();
        log = ""; // Inicializar el registro de eventos
        // Crear y añadir jugadores
        for (int i = 0; i < nplayers; i++) {
            char playerNumber = (char) ('0' + i);
            players.add(new Player(playerNumber, Dice.randomIntelligence(), Dice.randomStrength()));
        }
        // Determinar quién comienza y establecer el jugador actual
        currentPlayerIndex = Dice.whoStarts(nplayers);
        currentPlayer = players.get(currentPlayerIndex);
        configureLabyrinth();
        labyrinth.spreadPlayers(players);
    }

    public boolean finished(){
        return labyrinth.haveAWinner();
    }

    public boolean nextStep(Directions preferredDirections){
        boolean dead = this.currentPlayer.dead();
        if (!dead) {
            Directions direction = this.actualDirection(preferredDirections);

            if (direction != preferredDirections) {
                this.logPlayerNoOrders();
            }

            Monster monster = this.labyrinth.putPlayer(direction, currentPlayer);

            if (monster == null) {
                this.logNoMonster();
            } else {
                GameCharacter winner = this.combat(monster);
                this.manageReward(winner);
            }
        }
        else {
            this.manageResurrection();
        }

        boolean endGame = this.finished();

        if (!endGame) {
            this.nextPlayer();
        }

        return endGame;
    }

    public GameState getGameState(){
        return new GameState(labyrinth.toString(),players.toString(),monsters.toString(),currentPlayer.getNumber(),finished(),log);
    }

    private void configureLabyrinth(){
        Monster m1=new Monster("Monster 1", Dice.randomIntelligence(),Dice.randomStrength());
        labyrinth.addBlock(Orientation.HORIZONTAL, 2, 1, 1);
        labyrinth.addBlock(Orientation.HORIZONTAL, 2, 2, 1);
        labyrinth.addBlock(Orientation.VERTICAL, 1, 4, 1);
        labyrinth.addMonster(0,1, m1);
        monsters.add(m1);
    }

    private void nextPlayer(){
        // Incrementar el índice del jugador actual para que esté en rango
        currentPlayerIndex++;

        // Si el índice supera el último jugador, volver al primer jugador
        if (currentPlayerIndex >= players.size()) {
            currentPlayerIndex = 0;
        }

        // Actualizar el jugador actual
        currentPlayer = players.get(currentPlayerIndex);
    }

    private Directions actualDirection(Directions preferredDirection){
        int currentRow = this.currentPlayer.getRow();
        int currentCol = this.currentPlayer.getCol();
        Directions[] validMoves = labyrinth.validMoves(currentRow, currentCol).toArray(new Directions[0]);

        return this.currentPlayer.move(preferredDirection, validMoves); 
    }

    private GameCharacter combat(Monster monster){
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;

        float playerAttack = this.currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);

        while ((!lose) && (rounds < MAX_ROUNDS)){
            winner = GameCharacter.MONSTER;
            rounds++;

            float monsterAttack = monster.attack();
            lose = this.currentPlayer.defend(monsterAttack);

            if (!lose) {
                playerAttack = this.currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }

        this.logRounds(rounds, MAX_ROUNDS);

        return winner;
    }

    private void manageReward(GameCharacter winner){
        if (winner == GameCharacter.PLAYER){
            this.currentPlayer.receiveReward();
            this.logPlayerWon();
        }
        else{
            this.logMonsterWon();
        }
    }

    private void manageResurrection(){
        boolean resurrect = Dice.resurrectPlayer();

        if (resurrect){
            this.currentPlayer.resurrect();
            this.logResurrected();
        }
        else{
            this.logPlayerSkipTurn();
        }
    }

    private void logPlayerWon(){
        log += currentPlayer.getName() + " ha ganado el combate.\n";
    }

    private void logMonsterWon(){
        log += "El monstruo ha ganado el combate.\n";
    }

    private void logResurrected(){
        log += currentPlayer.getName() + " ha resucitado.\n";
    }

    private void logPlayerSkipTurn(){
        log += currentPlayer.getName() + " ha perdido el turno por estar muerto.\n";
    }

    private void logPlayerNoOrders(){
        log += currentPlayer.getName() + " no ha seguido las instrucciones del jugador humano (no fue posible).\n";
    }

    private void logNoMonster(){
        log += currentPlayer.getName() + " se ha movido a una celda vacía o no le ha sido posible moverse.\n";
    }

    private void logRounds(int rounds, int max){
        log += "Se han producido " + rounds + " de " + max + " rondas de combate.\n";
    }
}

File: irrgarten/Game.rb
module Irrgarten
  require_relative 'Dice'
  require_relative 'Labyrinth'
  require_relative 'Player'
  require_relative 'GameState'
  require_relative 'Monster'
  require_relative 'Orientation'
  require_relative 'GameCharacter'
  class Game
    @@MAX_ROUNDS=10

    #COnstructor
    def initialize(nplayers)
      @log = ""

      @monsters=Array.new
      @players=Array.new
      nplayers.times do |i|
        @players<< Player.new(i, Dice.randomIntelligence, Dice.randomStrength)
      end
      #Decides quien empieza y lo asignas a current player
      @currentPlayerIndex=Dice.whoStarts(nplayers)
      @currentPlayer=@players[@currentPlayerIndex]
      #Creas un nuevo laberinto
      # weapon = Weapon.new(Dice.weaponPower, Dice.usesLeft)
      @labyrinth = Labyrinth.new(10,10,9,9)
      configureLabyrinth
      #por implementar
      @labyrinth.spreadPlayers(@players)
    end
    #Si el juego ha terminado
    def finished
      return @labyrinth.haveAWinner
    end
    #POr implementar
    def nextStep(preferredDirection)
      @log=""
      dead=@currentPlayer.dead
      if !dead
        direction=self.actualDirection(preferredDirection)
        if direction != preferredDirection
          self.logPLayerNoOrders
        end
        monster=@labyrinth.putPlayer(direction,@currentPlayer)
        if monster==nil
          self.logNoMonster
        else
          winner=combat(monster)
          manageReward(winner)
        end
      else
        manageResurrection
      end
      endGame=finished
      if !endGame
        self.nextPlayer
      end
      return endGame
    end
    #Da el gamestate
    def getGameState
      p = ""
      for i in 0..@players.size-1
        p+=@players[i].to_s + "\n"
      end
      m= ""
      for i in 0..@monsters.size-1
        m+=@monsters[i].to_s + "\n"
      end
      laberinto = @labyrinth.to_s
      return GameState.new(laberinto,p,m,@currentPlayerIndex,finished,@log)
    end
    private
    def configureLabyrinth
      @labyrinth.addBlock(Orientation::HORIZONTAL, 0, 5, 4)
      @labyrinth.addBlock(Orientation::VERTICAL, 1, 1, 4)
      @labyrinth.addBlock(Orientation::VERTICAL, 2, 3, 4)
      @labyrinth.addBlock(Orientation::VERTICAL, 3, 8, 3)
      @labyrinth.addBlock(Orientation::VERTICAL, 7, 1, 2)
      @labyrinth.addBlock(Orientation::VERTICAL, 7, 3, 2)
      @labyrinth.addBlock(Orientation::VERTICAL, 8, 5, 3)
      #Cambiar intelligence y strengh
      for i in 1..6
        @monsters<< Monster.new("Monstruo #{i}",Dice.randomIntelligence,Dice.randomStrength)
      end
      @labyrinth.addMonster(0, 3, @monsters[0])
      @labyrinth.addMonster(0, 9, @monsters[1])
      @labyrinth.addMonster(2, 7, @monsters[2])
      @labyrinth.addMonster(4, 9, @monsters[3])
      @labyrinth.addMonster(7, 2, @monsters[4])
      @labyrinth.addMonster(7, 6, @monsters[5])
      @labyrinth.addMonster(9, 6, @monsters[6])
    end
    #Pasa al siguiente jugador
    def nextPlayer
      #puts @currentPlayerIndex
      c = @currentPlayerIndex+1
      @currentPlayerIndex = c % @players.size
      #puts @currentPlayerIndex
      @currentPlayer = @players[@currentPlayerIndex]
    end
    #Por implementar
    def actualDirection(preferredDirection)
      currentRow=@currentPlayer.getRow
      currentCol=@currentPlayer.getCol
      validmoves=@labyrinth.validMoves(currentRow,currentCol)
      output=@currentPlayer.move(preferredDirection,validmoves)
      return output
    end
    #Por implementar
    def combat(monster)
      rounds=0
      winner=GameCharacter::PLAYER
      playerAttack=@currentPlayer.attack
      lose=monster.defend(playerAttack)
      while (!lose && rounds<=@@MAX_ROUNDS)
        winner=GameCharacter::MONSTER
        rounds+=1
        monsterAttack=monster.attack
        lose=@currentPlayer.defend(monsterAttack)
        if !lose
          playerAttack=@currentPlayer.attack
          winner=GameCharacter::PLAYER
          lose=monster.defend(playerAttack)
        end
        self.logRounds(rounds,@@MAX_ROUNDS)
        return winner

      end
    end
    #Por implementar
    def manageReward(winner)
      if winner == GameCharacter::PLAYER
        @currentPlayer.receiveReward
        self.logPlayerWon
      else
        self.logPLayerSkipTurn
      end
    end
    #Por implementar
    def manageResurrection
      resurrect= Dice.resurrectPlayer
      if resurrect
        @currentPlayer.resurrect
        self.logResurrected
      else
        self.logPLayerSkipTurn
      end
    end
    #Por implementar
    def logPlayerWon
      @log += "El jugador ha ganado el combate.\n"
      puts @log
    end
    def logMonsterWon
      @log += "El monstruo ha ganado el combate.\n"
      #puts @log
    end
    def logResurrected
      @log += "El jugador #{@current_player_index} ha resucitado.\n"
      #puts @log
    end
    def logPLayerSkipTurn
      @log += "El jugador #{@current_player_index}  perdido el turno por estas muerto.\n"
      #puts @log
    end
    def logPLayerNoOrders
      @log += "El jugador #{@current_player_index}  no ha seguido las instrucciones indicadas.\n"
      #puts @log
    end
    def logNoMonster
      @log += "El jugador #{@current_player_index}  ha llegado a una casilla vacia o no ha podido moverse.\n"
    end
    def logRounds(rounds,max)
      @log += "Nos encontramos en la ronda #{rounds} de #{max} rondas. \n"
    end

  end
end